src/validator.rs

```diff
@@ -188,39 +246,50 @@ impl<T> InquireLength for &[T] {
 
 /// Built-in validator that checks whether the answer is not empty.
 ///
-/// # Arguments
-///
-/// * `$message` - optional - Error message returned by the validator.
-///   Defaults to "A response is required."
-///
 /// # Examples
 ///
 /// ```
-/// use inquire::{required, validator::{StringValidator, Validation}};
+/// use inquire::validator::{StringValidator, Validation, ValueRequiredValidator};
 ///
-/// let validator: StringValidator = required!();
-/// assert_eq!(Validation::Valid, validator("Generic input")?);
-/// assert_eq!(Validation::Invalid("A response is required.".into()), validator("")?);
+/// let validator = ValueRequiredValidator::new();
+/// assert_eq!(Validation::Valid, validator.validate("Generic input")?);
+/// assert_eq!(Validation::Invalid("A response is required.".into()), validator.validate("")?);
 ///
-/// let validator: StringValidator = required!("No empty!");
-/// assert_eq!(Validation::Valid, validator("Generic input")?);
-/// assert_eq!(Validation::Invalid("No empty!".into()), validator("")?);
+/// let validator = ValueRequiredValidator::new().with_message("No empty!");
+/// assert_eq!(Validation::Valid, validator.validate("Generic input")?);
+/// assert_eq!(Validation::Invalid("No empty!".into()), validator.validate("")?);
 /// # Ok::<(), inquire::error::CustomUserError>(())
 /// ```
-#[macro_export]
-#[cfg(feature = "builtin_validators")]
-macro_rules! required {
-    () => {
-        $crate::required! {"A response is required."}
-    };
+#[derive(Clone, Default)]
+pub struct ValueRequiredValidator {
+    message: Option<String>,
+}
 
-    ($message:expr) => {{
-        use $crate::validator::{ErrorMessage, Validation};
-        &|a| match a.is_empty() {
-            false => Ok(Validation::Valid),
-            true => Ok(Validation::Invalid(String::from($message).into())),
-        }
-    }};
+impl ValueRequiredValidator {
+    /// Create a new instance of this validator with default error message.
+    pub fn new() -> Self {
+        Self { message: None }
+    }
+
+    /// Define a custom error message returned by the validator.
+    /// Defaults to `A response is required.`.
+    pub fn with_message(mut self, message: impl Into<String>) -> Self {
+        self.message = Some(message.into());
+        self
+    }
+}
+
```

Comment:
What do you think about the other validators? I believe it may be reasonable to do the same there.

For example, the `MaxLengthValidator` could become:
```rust
#[derive(Clone)]
pub struct MaxLengthValidator {
    limit: usize,
    message: String,
}

impl MaxLengthValidator {
    /// Create a new instance of this validator, requiring at most the given length, otherwise
    /// returning an error with given message.
    ///
    /// If the message is omitted, `The length of the response should be at most $length` is used
    /// as default.
    pub fn new(limit: usize, message: Option<impl Into<String>>) -> Self {
        Self {
            limit,
            message: match message {
                Some(msg) => msg.into(),
                None => format!("The length of the response should be at most {}", limit),
            },
        }
    }
}
```

Reply:
Actually, better to not do that. If we pass `None` for `message`, then rustc can't infer the type and would require a very _ugly_ function call, where all we want is, simply passing `None` as argument.